pallets/auctions/src/types/english.rs

```diff
@@ -0,0 +1,118 @@
+use super::*;
+
+///
+/// Implementation of EnglishAuction
+///
+impl<T: Config> NftAuction<T::AccountId, T::AuctionId, BalanceOf<T>, Auction<T>, Bid<T>> for EnglishAuction<T> {
+	fn create(&self, sender: T::AccountId, auction: &Auction<T>) -> DispatchResult {
+		self.validate_data()?;
+		Pallet::<T>::validate_create(&self.common_data)?;
+		Pallet::<T>::handle_create(sender, auction, &self.common_data)?;
+
+		Ok(())
+	}
+
+	fn update(self, sender: T::AccountId, auction_id: T::AuctionId) -> DispatchResult {
+		self.validate_data()?;
+
+		<Auctions<T>>::try_mutate(auction_id, |maybe_auction| -> DispatchResult {
+			let auction_result = maybe_auction.as_mut().ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+			if let Auction::English(english_auction) = auction_result {
+				Pallet::<T>::validate_update(sender, &english_auction.common_data)?;
+				*english_auction = self;
+
+				Ok(())
+			} else {
+				Err(Error::<T>::NoChangeOfAuctionType.into())
+			}
+		})
+	}
+
+	///
+	/// Places a bid on an EnglishAuction
+	///
+	/// - removes lock on auction.common_data.last_bid
+	/// - sets lock on new bid
+	/// - updates auction.common_data.last_bid and auction.common_data.next_bid_min
+	/// - if necessary, increases auction end time to prevent sniping
+	///
+	fn bid(&mut self, _auction_id: T::AuctionId, bidder: T::AccountId, bid: &Bid<T>) -> DispatchResult {
+		// Lock / Unlock funds
+		if let Some(current_bid) = &self.common_data.last_bid {
+			<T as crate::Config>::Currency::remove_lock(AUCTION_LOCK_ID, &current_bid.0);
+		}
+		<T as crate::Config>::Currency::set_lock(AUCTION_LOCK_ID, &bidder, bid.amount, WithdrawReasons::all());
+
+		self.common_data.last_bid = Some((bidder, bid.amount));
+		// Set next minimal bid
+		Pallet::<T>::set_next_bid_min(&mut self.common_data, bid.amount)?;
+
+		// Avoid auction sniping
+		Pallet::<T>::avoid_auction_sniping(&mut self.common_data)?;
```

Comment:
Both of these are fallible and write to storage -> not atomic without `#[transactional]` annotation on the extrinsic